Scrollable control areas of widgets

markotoplak opened the pull request:
##### Issue
Eventually attempts to fix #3058.

##### Includes
- [ ] Code changes
- [X] Tests
- [ ] Documentation

astaric commented on Orange/widgets/tests/base.py:
```diff
@@ -364,6 +364,26 @@ def modifiers(self, modifiers):
+    def test_minimum_size(self):
+        widget = getattr(self, "widget", None)
+        if widget is None:
+            return
```
Or skip the tests using self.skipTest("minimum size needs to be tested manually")?

Otherwise, it will be marked as passing, even if it does not do anything.

astaric commented on Orange/widgets/tests/base.py:
```diff
@@ -364,6 +364,26 @@ def modifiers(self, modifiers):
+
+        invalidate_cached_size_hint(widget)
+        min_size = widget.minimumSizeHint()
+        # FIXME high-dpi screens
```
On mac, size returns size in virtual, not device pixels, so the values you have are ok.

markotoplak commented:
@astaric, thanks.

There is a problem with adding tests to the base class: because of the import, they are also run separately with every test file. And ideas how to prevent this? 

Before I did not even notice it because of the return. :)

astaric commented:
As far as I know, there are no simple tricks you can use.

The proper way would probably be a mixin that contains just the test methods, but maybe something could be done with metaclasses? A metaclass that removes the method when the baseclass is instantiated, but remembers in somewhere so it can put it back for the subclasses? :)

astaric commented:
Something like this :)
```
class MyMeta(type(GuiTest)):
    def __new__(cls, name, bases, namespace):

        if name == "WidgetTest":
            cls.tests = {}
            for key in list(namespace):
                if key.startswith("test"):
                    cls.tests[key] = namespace.pop(key)
        else:
            for key, method in cls.tests.items():
                namespace[key] = method

        obj = super().__new__(cls, name, bases, namespace)
        return obj

class WidgetTest(GuiTest, metaclass=MyMeta):
```

janezd commented:
Metaclasses probably work, but `__init_subclass__` may be more appropriate because this is something that `WidgetTest` wants to do to its subclasses.

```
class A:
    def __init_subclass__(cls):
        def f(self, x):
            return self.k * x
        cls.f = f

class B(A):
    def __init__(self, k):
        self.k = k

b = B(2)
print(b.f(3))  # prints 6


a = A()
print(a.f(3))  # error
```

markotoplak commented:
To make large widgets work with smaller screens, I added scroll areas to widgets' left sides. 

For now, to highlight potential issues, I added it to all widgets. Could anyone play around with this?

I added size hints so that scroll bars should not appear if you add a new widget, but if a user saved a workflow on a Mac (which has quite a compact interface) and opened on Ubuntu (with lots of empty space), I think the scroll bars would appear as window sizes are saved (didn't try this yet, though).

Therefore I recommend adding an attribute to OWWidget that controls if scroll panes are used for a certain widget. Thoughts? @lanzagar, @janezd, @ales-erjavec

lanzagar commented:
The scrollbar looks and functions well on my system.

I agree with this being optional and only manually turned on for certain widgets to reduce unwanted side effects like size changes for saved workflows on different systems.

One downside currently is that our (separate) very common problem of bad minimum widget size settings becomes even more evident. For example Scatter plot has a long standing issue where the main area can be reduced too much and looks bad when that sometimes happens (unintentionally). Control area was at least enforcing a vertical minimum size and now even that might be gone.
So together with this, I think we should set some minimum sizes, at least for widgets where scrolling will be turned on (but wouldn't hurt in some other cases too).

janezd commented:
It's unfortunate that macOS shows scroll bars only when scrolling (unless the user changes the default preference). This causes the visualization to resize. But we'll live with that.

I'd prefer moving code out of the huge Orange/widgets/widget.py. But gui.py is also one large mess. If you think this code will be useful elsewhere, please move it to gui.py. If you won't, then say, and we can merge.

ajdapretnar commented:
For reference: go to System Preferences - General - Show scroll bars.

thocevar commented:
There's a problem with hiding the control area. If the scrollbar is present and you hide and reopen control area, the sizes compute incorrectly and the scrollbar overlaps some content.

janezd commented:
Indeed, this also happens on macOS, if you set Show Scroll Bars in system preferences to Always.

markotoplak commented:
@thocevar, @janezd, thank you for the wrong resize problem if it is collapsed. It can reproduce it easily on a mac. I though I had seen it before and also fixed it while I was experimenting.

markotoplak commented:
I think that the problem with wrong size after control area show is not fixed.

As @janezd mentions, scroll bars, when hidden as on a Mac if you are not using a mouse :), should not resize the window, but I have no idea how to implement that properly.

I moved some reusable code to gui and the other class to inside OWWidget. I think this is ready to be merged.
